doc/website/blog/2021-11-1-sync-objects-in-external-database.md

```diff
@@ -0,0 +1,305 @@
+go run entgo.io/ent/cmd/ent init User
+```
+
+Then, add the `name` and the `avatar_url` fields and run `go generate` for generating the assets.
+
+```go title="ent/schema/user.go"
+type User struct {
+	ent.Schema
+}
+
+func (User) Fields() []ent.Field {
+	return []ent.Field{
+		field.String("name").
+			Immutable(),
+		field.String("avatar_url").
+			Immutable(),
+	}
+}
+```
+
+```shell
+go generate ./ent
+```
+
+### The Problem
+
+The `avatar_url` field defines a URL to an image in a bucket on our object storage (e.g. AWS S3). For the purpose of
+this discussion  we want to make sure that:
+
+- When a user is created, an image with the URL stored in `"avatar_url"` exists in our bucket.
+- Orphan images are deleted from the bucket. This means that when a user is deleted from our system, its avatar image
+  is deleted as well.
+  
+For interacting with blobs, we will use the [`gocloud.dev/blob`](https://gocloud.dev/howto/blob) package. This package
+provides abstraction for reading, writing, deleting and listing blobs in a bucket. Similar to the `database/sql`
+package, it allows interacting with variety of object storages with the same API by configuring its driver URL. 
+For example:
+
+```go
+// Open an in-memory bucket. 
+if bucket, err := blob.OpenBucket(ctx, "mem://photos/"); err != nil {
+	log.Fatal("failed opening in-memory bucket:", err)
+}
+
+// Open an S3 bucket named photos.
+if bucket, err := blob.OpenBucket(ctx, "s3://photos"); err != nil {
+	log.Fatal("failed opening s3 bucket:", err)
+}
+
+// Open a bucket named photos in Google Cloud Storage.
+if bucket, err := blob.OpenBucket(ctx, "gs://my-bucket"); err != nil {
+	log.Fatal("failed opening gs bucket:", err)
+}
+```
+
+### Schema Hooks
+
+[Hooks](https://entgo.io/docs/hooks) are a powerful feature of Ent that allows adding custom logic before and after
+operations that mutate the graph.
+
+Hooks can be either defined dynamically using `client.Use` (called "Runtime-hooks"), or explicitly on the schema
+(called "Schema Hooks") as follows:
+
+```go
+// Hooks of the User.
+func (User) Hooks() []ent.Hook {
+	return []ent.Hook{
+		EnsureImageExists(),
+		DeleteOrphans(),
+	}
+}
+```
+
+As you can imagine, the `EnsureImageExists` hook will be responsible for ensuring that when a user is created, their
+avatar URL exists in the bucket, and the `DeleteOrphans` will ensure that orphan images are deleted. Let's start
+writing them.
+
+```go title="ent/schema/hooks.go"
+func EnsureImageExists() ent.Hook {
+	hk := func(next ent.Mutator) ent.Mutator {
+		return hook.UserFunc(func(ctx context.Context, m *ent.UserMutation) (ent.Value, error) {
+			avatarURL, exists := m.AvatarURL()
+			if !exists {
+				return nil, errors.New("avatar field is missing")
+			}
+			// TODO:
+			// 1. Verify that "avatarURL" points to a real object in the bucket.
+			// 2. Otherwise, fail.
+			return next.Mutate(ctx, m)
+		})
+	}
+	// Limit the hook only to "Create" operations.
+	return hook.On(hk, ent.OpCreate)
```

Comment:
Nice, in the examples usually the condition for the hook is on the schema (and not on the hook) but that's just semantics.

Reply:
👍